### ISSUE TITLE
Referral promo support (& Dow Jones)

### ISSUE THREAD

**Issue by bbondy** [brave/brave-browser#287]
## Test plan
See https://github.com/brave/brave-core/pull/428

## Description
See below text

**Comment by bsclifton** [comment]
Both the macOS and Windows installers have the referral code in the executable name (instead of DMG on macOS we use PKG). The installer then strips that out and puts into the user's profile directory

The Windows implementation uses our fork of electron-winstaller (and the changes aren't very readable... since they involve binary file updates). The changes made for macOS are a little more interesting:
1. Use pkg installer for macOS in addition to of dmg image
    https://github.com/brave/browser-laptop/pull/12946
2. macOS: sign pkg installer
    https://github.com/brave/browser-laptop/pull/13033
3. mac - set file mode and group owner for pkg-installed Brave.app
    https://github.com/brave/browser-laptop/pull/13111
4. Update pkg-installed app permissions to $USER:admin
    https://github.com/brave/browser-laptop/pull/13132

Here's a link to an internal security review which has a link to a detailed spec (includes API details):
https://github.com/brave/internal/issues/250

Relevant PRs from browser-laptop showing the implementation:
1. Adds promo
    https://github.com/brave/browser-laptop/pull/13667
    Initial promo support

2. Adds referral code read into browser 
    https://github.com/brave/browser-laptop/pull/12798
    - Basically looks for a file `promoCode` in the user's profile folder
        - if present, reads it in, stores in state, and deletes the file
        - browser-laptop update service is hit (HTTP PUT) - `/promo/initialize/nonua` with promoCode
        - update service returns a unique downloadId which is then stored
    - If 30 days have passed, browser-laptop update service is hit again (passing downloadId)

3. Adds s-s to the referral code
    https://github.com/brave/browser-laptop/pull/13756
    Adds support for more use-cases than 3letter/3number

4. Check userAgent before trying to parse
    https://github.com/brave/browser-laptop/pull/13131
    We had a flaw in above logic (since partners can be sent the Brave UA) which caused a crash

**Comment by emerick** [comment]
@bsclifton Working on this now. Belated thanks on the great details!

**Comment by kjozwiak** [comment]
Win 10 x64 - **PASSED**

* installed using `BraveBrowserSetup-KAM253.exe`
* ensured that `promoCode` was being created under `C:\Users\kjozwiak\AppData\Local\BraveSoftware\Brave-Browser\User Data`
* waited 10-30s and ensured `promoCode` was removed from `C:\Users\kjozwiak\AppData\Local\BraveSoftware\Brave-Browser\User Data`
* checked `Local State` for `{"brave":{"referral":{"download_id":"75cfc36d-8a8d-4883-bda5-adcc1b5aa1a4","promo_code":"KAM253"}}`

Win 10 x86 - **PASSED**

* installed using `BraveBrowserSetup-86-KAM253.exe`
* ensured that `promoCode` was being created under `C:\Users\kjozwiakBrave\AppData\Local\BraveSoftware\Brave-Browser\User Data\`
* waited 10-30s and ensured `promoCode` was removed from `C:\Users\kjozwiakBrave\AppData\Local\BraveSoftware\Brave-Browser\User Data\`
* checked `Local State` for `{"brave":{"referral":{"download_id":"9d3aa2c6-8a5b-4c27-9aa2-67979c8888cd","promo_code":"KAM253"}}`

Win 8.1 x64 - **PASSED**

* installed using `BraveBrowserSetup-KAM253.exe`
* ensured that `promoCode` was being created under `C:\Users\kamil\AppData\Local\BraveSoftware\Brave-Browser\User Data\`
* waited 10-30s and ensured `promoCode` was removed from `C:\Users\kamil\AppData\Local\BraveSoftware\Brave-Browser\User Data\`
* checked `Local State` for `{"brave":{"referral":{"download_id":"c8da7daa-e980-44c4-a3f0-1335df6cb9fd","promo_code":"KAM253"}}`

Win 7 x86 - **PASSED**

* installed using `BraveBrowserSetup-86-KAM253.exe`
* ensured that `promoCode` was being created under `C:\Users\kjozwiak\AppData\Local\BraveSoftware\Brave-Browser\User Data\`
* waited 10-30s and ensured `promoCode` was removed from C:\Users\kjozwiak\AppData\Local\BraveSoftware\Brave-Browser\User Data\`
* checked `Local State` for `{"brave":{"referral":{"download_id":"896026da-4aae-495c-8b7e-32c93d7af4b1","promo_code":"KAM253"}}`

macOS 10.13.6 x64 - **PASSED**

* installed using `Brave-Browser-KAM253.pkg`
* ensured that `promoCode` was being created under `~/Library/Application\ Support/BraveSoftware/Brave-Browser/`
* launched Brave
* waited 10-30s and ensured `promoCode` was removed from `~/Library/Application\ Support/BraveSoftware/Brave-Browser/`
* checked `Local State` for `{"brave":{"referral":{"download_id":"fe1126e4-58cc-435a-a5ef-043927eec8d0","promo_code":"KAM253"}}`

**Comment by kjozwiak** [comment]
Once the above test were done, @aekeus sent QA a list of ID's associated with `KAM253` and we ensured that the above `download_id` were also being listed.

**Comment by kjozwiak** [comment]
QA we'll need to work with @emerick to figure out how we can check the following using b-c:

• Verifying that we no longer send referral code to update server after 90 days
• Verifying that that service checks for referral finalization 30 days after referral initialization
• Verifying that service retries checking for referral finalization up to 30 times (with 24-hour pauses between attempts) if it doesn't receive "true" on its first try.

**Comment by kjozwiak** [comment]
Went through verification using the following build under `macOS 10.14 x64` - **PASSED**

Brave | 0.56.12 Chromium: 70.0.3538.77 (Official Build) (64-bit)
-- | --
Revision | 0f6ce0b0cd63a12cb4eccea3637b1bc9a29148d9-refs/branch-heads/3538@{#1039}
OS | Mac OS X

* launch brave via `BRAVE_REFERRALS_SERVER=laptop-updates-staging.herokuapp.com` (make sure you're using a referral code that's on stating and NOT production)
* wait 10min (this is what the staging server is set too) or you'll receive `Referral is not ready, please wait at least 30 days` if you do the below steps without waiting 10min
* after waiting 10min, relaunch using `BRAVE_REFERRALS_SERVER=laptop-updates-staging.herokuapp.com BRAVE_REFERRALS_CHECK_TIME=5`
* you should see a `timestamp` being created next to the `promo_code`, example: `"promo_code":"MIM486","timestamp":"13186203834926030"`
